mars/dataframe/execution/arithmetic.py

```diff
@@ -82,15 +83,15 @@ def _index_align_map(ctx, chunk):
 
 
 def _index_align_reduce(ctx, chunk):
-    input_idx_to_df = {inp.index: ctx[inp.key, ','.join(str(idx) for idx in chunk.index)]
-                       for inp in chunk.inputs[0].inputs}
-    row_idxes = sorted({idx[0] for idx in input_idx_to_df})
-    col_idxes = sorted({idx[1] for idx in input_idx_to_df})
+    input_keys = get_shuffle_reduce_inputs(chunk.inputs[0])
+    idxes_gen = itertools.product(range(chunk.op.row_source_size), range(chunk.op.col_source_size))
```

Comment:
it's indeed a little bizarre to  handle all chunk inputs' indexes in such a way, I recommend to add a `_to_fetch_idxes` to `FetchShuffle`, and `row_source_size` and `col_source_size` could not be required any more.